[Yuvaraj Kumaresan] iP

Yuvaraj0702 opened the pull request:
# Duke the chatbot
> "Why do you even need to remember." - Yuvaraj Kumaresan

Duke makes sure you never need to remember anything ever. It is,
- text-based
- easy to learn
- ~~Fast~~ Fastest to use.

All you need to do is,
1. download it from [here](https://github.com/nus-cs2103-AY2223S1/ip/pull/28)
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grinning:

And it is **FREE!**

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

___
If you Java programmer, you can use it to practice Java too. Here's the main method:

```
public class Main {
  public static void main(String[] args) throws DukeException, IOException {
        String divider = "---------------------------------------------------" +
                "-------------------------------------------------------------";

        String logo = " ____        _        \n"
                + "|  _ \\ _   _| | _____ \n"
                + "| | | | | | | |/ / _ \\\n"
                + "| |_| | |_| |   <  __/\n"
                + "|____/ \\__,_|_|\\_\\___|\n";
        System.out.println(divider);
        System.out.println("Hello I am \n" + logo + "\nType help to see user guide.\n\nWhat can I do for you?");
        System.out.println(divider);
        run();
    }
}
```

Yuvaraj0702 committed:
Level one fearures added

Yuvaraj0702 committed:
Merge branch 'master' of https://github.com/Yuvaraj0702/ip

Yuvaraj0702 committed:
Level 1 features have been added to Duke

Yuvaraj0702 committed:
Level 1 features added

Yuvaraj0702 committed:
Level 1 features added

Yuvaraj0702 committed:
Intermediate version of level 2

Yuvaraj0702 committed:
Correct implementation for level 2

Yuvaraj0702 committed:
Level 2 features implemented

Yuvaraj0702 committed:
Intermediary commit of level 3 features

Yuvaraj0702 committed:
Bug fixes to include indices more than 9

Yuvaraj0702 committed:
Implemented level 3 features

Yuvaraj0702 committed:
Adding pre requisites for level 4

Yuvaraj0702 committed:
Intermediary version of level 4 features

Yuvaraj0702 committed:
Implemented level 4 features

Yuvaraj0702 committed:
Copied Duke.java over to src folder for testing and tried testing

Yuvaraj0702 committed:
fixing file structure issues

Yuvaraj0702 committed:
Fixing file structure issues part 2

Yuvaraj0702 committed:
Automated testing of text Ui completed

Yuvaraj0702 committed:
File structure improvements

Yuvaraj0702 committed:
Satisfactorily implemented level 5 features

Yuvaraj0702 committed:
Level 6 features implemented

Yuvaraj0702 committed:
Cosmetic changes

Yuvaraj0702 committed:
Edits to javadocs

Yuvaraj0702 committed:
Bug fixes

Yuvaraj0702 committed:
Add basic functionality for saving to file

Yuvaraj0702 committed:
Change output format

Yuvaraj0702 committed:
Implement level 7 features

Yuvaraj0702 committed:
Add preliminary level 8 features

Yuvaraj0702 committed:
Add different formatting for date time output

Yuvaraj0702 committed:
Merge branch 'branch-Level-7'

Yuvaraj0702 committed:
Merge branch 'branch-Level-8' of https://github.com/Yuvaraj0702/ip

Yuvaraj0702 committed:
Post merge commit

Yuvaraj0702 committed:
Post merge commit

Yuvaraj0702 committed:
Fix post merge issues

Yuvaraj0702 committed:
Fix Level-8 merge issues

Yuvaraj0702 committed:
Reorganize file structure

Yuvaraj0702 committed:
Add A-MoreOOP features

Yuvaraj0702 committed:
Add A-Packages features

Yuvaraj0702 committed:
Add A-Gradle functionality

Yuvaraj0702 committed:
Bug fixes

Yuvaraj0702 committed:
Fix more bugs

Yuvaraj0702 committed:
Add A-JUnit features

Yuvaraj0702 committed:
Add javadoc comments

Yuvaraj0702 committed:
Change to fit coding standard

Yuvaraj0702 committed:
Add branch levl 9 features

Yuvaraj0702 committed:
Merge branch 'branch-A-JavaDoc'

Yuvaraj0702 committed:
Merge branch 'branch-A-CodingStandard'

Yuvaraj0702 committed:
Merge branch 'branch-Level-9'

Yuvaraj0702 committed:
Add style changes

Yuvaraj0702 committed:
Update README.md

Yuvaraj0702 committed:
Change JavaDocs

Yuvaraj0702 committed:
Change JavaDoc headers

Yuvaraj0702 committed:
Fix style and punctuation

Yuvaraj0702 committed:
Add author tag

Yuvaraj0702 committed:
Add missing full stops

Bahamas20 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,366 @@
+    /**
+     * Handles UI output for list keyword.
+     */
+    public void list() {
```
Perhaps you can change the naming to a verb instead ? For eg: 
```suggestion
    public void showList() {
```

Bahamas20 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,366 @@
+    /**
+     * Handles UI output for delete keyword with no line number.
+     */
+    public void deleteNoNumber() {
```
I was confused initially by the naming of this method, maybe you could change it to what is shown below instead ?
```suggestion
    public void deleteWithNoNumber() {
```

Bahamas20 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,366 @@
+    /**
+     * Handles the help keyword and displays the user guide.
+     */
+    public void help() {
```
This method is very well done and it was something that most people would have overlooked , Great Job 👏

Bahamas20 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,45 @@
+    /**
+     * Time at which the event is at.
+     */
+    protected String at;
```
Perhaps you could be more specific with the naming of the at variable as it may not be clear to others ?
```suggestion
    protected String eventAt;
```

zanwenfu replied:
It's better to name it as eventTime or time. Others might be confused what "at" means.

Bahamas20 commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+    /**
+     * Time at which the deadline is due.
+     */
+    protected LocalDateTime by;
```
Same comment as Event variable at.
```suggestion
    protected LocalDateTime deadlineBy;
```

Bahamas20 reviewed (commented):
Overall very well written and thorough code, naming for most variables were also well done.
Just a few nits to fix. 💯 👍

zanwenfu commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,47 @@
+package duke;
+
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
```
Good that you separate package and import declaration.

zanwenfu commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Duke {
+
+    private static TaskList taskList = new TaskList(new ArrayList<Task>());
+    private static Storage storage = new Storage(taskList);
+    private static Ui ui = new Ui(storage);
+    private static Parser parser = new Parser(storage, ui);
```
Good that you make class variables private.

zanwenfu commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,49 @@
+package duke;
+
+/**
+ * Handles possible exceptions thrown by duke.Duke.
+ *
+ * @author Yuvaraj Kumaresan
+ */
+
+public class DukeException extends Exception {
```
Class doc and class declaration should not be separated?

zanwenfu reviewed (commented):
Overall looks good!

Yuvaraj0702 committed:
Add level 10 features

Yuvaraj0702 committed:
Merge branch 'branch-Level-10'

CeereeC commented on src/main/java/duke/DukeException.java:
```diff
@@ -0,0 +1,49 @@
+    /**
+     * Handles possible errors with the duke.ToDo input phrase.
+     */
+    public static class DukeToDoException extends DukeException {
```
Good job modularising the exceptions beyond just DukeException

CeereeC commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+                String[] description = text.replace("deadline ", "").split("/by ");
+                try {
+                    if (description[1].length() > 10) {
+                        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm");
+                        Deadline item = new Deadline(description[0], LocalDateTime.parse(description[1], formatter));
+                        storage.taskList.addDeadline(item);
+                        return ui.deadline(item);
+                    } else {
+                        description[1] = description[1] + " 00:00";
+                        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd-MM-yyyy HH:mm");
+                        Deadline item = new Deadline(description[0], LocalDateTime.parse(description[1], formatter));
+                        storage.taskList.addDeadline(item);
+                        return ui.deadline(item);
```
Might be neater to abstract this into another sub-function due to the repetition

CeereeC commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,183 @@
+                return ui.generalError(error);
+            }
+        }
+    }
```
Perhaps using a switch statement could be neater? Also, given how you have many specific Duke Exceptions, you could consider moving all the catch blocks below, and capture the rest of the logic in one try block.

CeereeC commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,97 @@
+    public static void writeToFile() throws IOException {
+        FileWriter dukeWriter = new FileWriter("duke.txt");
+        for (int i = 0; i < taskList.arrayList.size(); i++) {
+            if (taskList.arrayList.get(i).getClass() == ToDo.class) {
```
An alternative is you could use is Enums as well

CeereeC commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,366 @@
+    /**
+     * Handles the help keyword and displays the user guide.
+     */
+    public String help() {
```
Very good that you included a help command

CeereeC reviewed (commented):
Overall, I found your code easy to read, though there are a few places where the nesting was a little too deep.

Yuvaraj0702 committed:
Add Assertions

This commit adds assertions to the duke chatbot.
It asserts to check if the images and text submitted are present.
Flags off if they are null.

Yuvaraj0702 committed:
Fix code quality

This commit fixes issues such as deep nesting,
arrow methods and long methods.
It also fixes some bugs that went unnoticed in previous commits.

Yuvaraj0702 committed:
Merge pull request #3 from Yuvaraj0702/branch-A-Assertions

Add Assertions

Yuvaraj0702 committed:
Merge branch 'master' into branch-A-CodeQuality

Yuvaraj0702 committed:
Merge pull request #2 from Yuvaraj0702/branch-A-CodeQuality

Fix code quality

Fixes issues such as deep nesting, long methods and arrow methods.
Some bug fixes are also included.

Yuvaraj0702 committed:
Add fixed duration functionality

Adds the ability to add a task that only has a estimated completion time.
This task has all the functionalities of other tasks as well.
Also includes bug fixes and style changes.

Yuvaraj0702 committed:
Merge pull request #4 from Yuvaraj0702/branch-B-FixedDurationTasks

Add fixed duration functionality

Yuvaraj0702 committed:
Add Ui image and README

Yuvaraj0702 committed:
Add more java docs and fix bugs

Yuvaraj0702 committed:
Add changes to allow for gradle run

Yuvaraj0702 committed:
Fix spelling errors

Yuvaraj0702 committed:
Add welcome message
